### PR TITLE
FIR checkers: report specific errors instead of INAPPLICABLE_CANDIDATE

### PR DESCRIPTION
**Pull request by tgeng**
Specifically, the report the following 4 errors.

 * NON_VARARG_SPREAD
 * ARGUMENT_PASSED_TWICE
 * TOO_MANY_ARGUMENTS
 * NO_VALUE_FOR_PARAMETER

In addition, to added NAME_OF_NAMED_ARGUMENT position strategy for some of the
new checks.

### CODE REVIEW

**Comment by tgeng** [comment]
Note that this change will need to be rebased on top of https://github.com/JetBrains/kotlin/pull/4172 in order to report multiple errors in the call. I will update all the test cases again after https://github.com/JetBrains/kotlin/pull/4172 is merged.

File: compiler/testData/diagnostics/tests/testsWithJava15/jvmRecord/simpleRecords.fir.kt
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/testsWithJava15/jvmRecord/simpleRecords.fir.kt:11]: This shouldn't be `TOO_MANY_ARGUMENTS`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/testsWithJava15/jvmRecord/simpleRecords.kt#L11

Comment (jsjeon) [reply compiler/testData/diagnostics/tests/testsWithJava15/jvmRecord/simpleRecords.fir.kt:11]: Ah, it was `INAPPLICABLE_CANDIDATE` anyway, so the question is why/where the root cause was set as `TooManyArguments`.

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt
Comment (jsjeon) [inline compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:176]: Not `KtReferenceExpression`?

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt
Comment (jsjeon) [inline compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:180]: Don't we need to introduce and use positioning strategy `VALUE_ARGUMENTS`?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/coneDiagnosticToFirDiagnostic.kt
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/coneDiagnosticToFirDiagnostic.kt:104]: Could you point out some changed examples?

File: compiler/testData/diagnostics/tests/checkArguments/kt1940.fir.kt
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/checkArguments/kt1940.fir.kt:10]: Nice! :)

File: compiler/testData/diagnostics/tests/deprecated/javaDeprecated.fir.kt
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/deprecated/javaDeprecated.fir.kt:14]: Did you investigate this case? Is it safe to remove those errors?

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt
Comment (tgeng) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:176]: So we talked about whether to report the entire argument and use a position strategy to pick up the name or report the name directly. Since we agreed to do the former I am reporting the entire argument. If we want to report the name, I will need to add a util function that locates the name part from an argument.

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt
Comment (tgeng) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:180]: Good point! Let me do that

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt
Comment (jsjeon) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:176]: I didn't ask about positioning; it was simply `s/KtValueArgument/KtReferenceExpression/g` (to match with old diagnostic). Does it affect positioning?

Comment (jsjeon) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:176]: Duh, I forgot you mentioned this internally. Sorry!

Comment (tgeng) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:176]: Np! Sounds like we are on the same page now ;)

File: compiler/fir/analysis-tests/testData/resolve/arguments/lambda.kt
Comment (punzki) [inline compiler/fir/analysis-tests/testData/resolve/arguments/lambda.kt:31]: What's causing the difference in marking for LT and PSI?

Comment (tgeng) [reply compiler/fir/analysis-tests/testData/resolve/arguments/lambda.kt:31]: I think it's due to me running the light tree tests but not the PSI tests. Hence, the test data is updated to only have the light tree position updated with the new position strategy. It should be back normal if I run the PSI test as well.

I will update all the test data later after all the needed dependency PRs are merged.

**Comment by tgeng** [comment]
I won't work on this PR for now until https://github.com/JetBrains/kotlin/pull/4186, which is blocked by https://youtrack.jetbrains.com/issue/KT-45344 is merged.

File: compiler/testData/diagnostics/tests/deprecated/javaDeprecated.fir.kt
Comment (tgeng) [reply compiler/testData/diagnostics/tests/deprecated/javaDeprecated.fir.kt:14]: This is safe since the more detailed error `NO_VALUE_FOR_PARAMETER` is disabled for this test case on top of this fiel

File: compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt
Comment (tgeng) [reply compiler/frontend/src/org/jetbrains/kotlin/diagnostics/PositioningStrategies.kt:478]: It would work and this way we will need to mark `NAMED_ARGUMENTS_NOT_ALLOWED` to be reported on `KtValueArgument` instead of `PsiElement`, which difffers from FE1.0. But I don't see a problem of doing this since our new checkers report on the whole argument rather than the name reference. Let me do it.

**Review by punzki — APPROVED** [review]

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually. FYI I had to add the following commit https://github.com/JetBrains/kotlin/commit/e7102cd63f861e3dd1472c708e524d1f8beac2c7, because your renderer built strange description string for "too many arguments" error.

**Pull request closed without merging**
